### PR TITLE
refactor(anta.tests)!: Update the pass/fail criteria for testcase VerifyBFD

### LINKED ISSUE #503: Update the pass/fail criteria for testcase VerifyBFD

### PR DESCRIPTION
**Pull request by MaheshGSLAB**
# Description

1. Moved VerifyBFD testcase from generic to bfd file and rename with VerifyBFDPeersHealth
2. Added the remote dick verification and an optional input parameter of last down to verify if peer was down more threshold hours.

Fixes #503 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have run pre-commit for code linting and typing (`pre-commit run`)
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

File: anta/tests/bfd.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+from datetime import datetime
+from typing import Optional
+
+from anta.models import AntaCommand, AntaTest
+
+
+class VerifyBFDPeersHealth(AntaTest):
+    """
+    Verifies there is no IPv4 BFD peer in the down state, remote disc is not zero and last down should be above the threshold for all VRF.
```
Comment (carl-baillargeon) [inline anta/tests/bfd.py:19]: What does `remote disc` mean?

File: anta/tests/bfd.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    # revision 1 as later revision introduces additional nesting for type
+    commands = [AntaCommand(command="show bfd peers", revision=1), AntaCommand(command="show clock")]
+
+    class Input(AntaTest.Input):
+        """
+        This class defines the input parameters of the testcase.
+        """
+
+        last_down: Optional[int] = None
```
Comment (carl-baillargeon) [inline anta/tests/bfd.py:37]:
```suggestion
        last_down: Optional[int] = Field(default=None, gt=0)
```

File: anta/tests/bfd.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+class VerifyBFDPeersHealth(AntaTest):
+    """
+    Verifies there is no IPv4 BFD peer in the down state, remote disc is not zero and last down should be above the threshold for all VRF.
+    Expected results:
+        * success: The test will pass if IPv4 BFD peers are not down, remote disc is not zero and last down above the defined threshold for all VRF.
+        * failure: The test will fail if IPv4 BFD peers are down, remote disc is zero and last down below the defined threshold for all VRF.
+    """
+
+    name = "VerifyBFDPeersHealth"
+    description = "Verifies there is no IPv4 BFD peer in the down state, remote disc is not zero and last down should be above the threshold for all VRF."
```
Comment (carl-baillargeon) [inline anta/tests/bfd.py:26]: I think we should mention that the last down threshold value is optional somehow.

File: anta/tests/bfd.py (hunk 2 above)
Comment (carl-baillargeon) [reply anta/tests/bfd.py:37]: Also, I am not sure I like `last_down` as an input. It is not very descriptive.

File: tests/units/anta_tests/test_bfd.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+            },
+            {
+                "utcTime": 1703658481.8778424,
+            },
+        ],
+        "inputs": {"last_down": 2},
+        "expected": {"result": "success"},
+    },
+    {
+        "name": "success-no-peer",
```
Comment (carl-baillargeon) [inline tests/units/anta_tests/test_bfd.py:64]: Shouldn't we fail the test if there is no peer?

File: anta/tests/bfd.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        # set the initial result
+        self.result.is_success()
+
+        # Iterate over IPv4 BFD peers
+        for vrf, vrf_data in bfd_output["vrfs"].items():
+            for peer, neighbor_data in vrf_data["ipv4Neighbors"].items():
+                for peer_data in neighbor_data["peerStats"].values():
+                    peer_status = peer_data["status"]
+                    remote_disc = peer_data["remoteDisc"]
+                    peer_l3intf = peer_data.get("l3intf", "")
```
Comment (carl-baillargeon) [inline anta/tests/bfd.py:60]: Do we need that `l3intf` key?

File: anta/tests/bfd.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        * success: The test will pass if IPv4 BFD peers are not down, remote disc is not zero and last down above the defined threshold for all VRF.
+        * failure: The test will fail if IPv4 BFD peers are down, remote disc is zero and last down below the defined threshold for all VRF.
+    """
+
+    name = "VerifyBFDPeersHealth"
+    description = "Verifies there is no IPv4 BFD peer in the down state, remote disc is not zero and last down should be above the threshold for all VRF."
+    categories = ["bfd"]
+
+    # revision 1 as later revision introduces additional nesting for type
+    commands = [AntaCommand(command="show bfd peers", revision=1), AntaCommand(command="show clock")]
```
Comment (carl-baillargeon) [inline anta/tests/bfd.py:30]: Can you check if it would be possible to refactor the test with an updated revision?

**Review by carl-baillargeon — CHANGES_REQUESTED** [review]

**Review by carl-baillargeon — COMMENTED** [review]
Thanks Mahesh. Since bfd.py is a new module, you need to create the documentation: anta/docs/api/tests.bfd.md and add the section in anta/docs/api/tests.md.

Finally add your newly created MD file to mkdocs.yml. Double check the doc with mkdocs serve.

Thanks

File: anta/tests/bfd.py (hunk 1 above)
Comment (MaheshGSLAB) [reply anta/tests/bfd.py:19]: is it ok to mention `discriminator value of the remote system`

File: anta/tests/bfd.py (hunk 3 above)
Comment (MaheshGSLAB) [reply anta/tests/bfd.py:26]: description = "Verifies there is no IPv4 BFD peer in the down state and discriminator value of the remote system is not zero for all VRF. BFD peer last down in hours is optional check which should be above the threshold for all VRF."

File: anta/tests/bfd.py (hunk 2 above)
Comment (MaheshGSLAB) [reply anta/tests/bfd.py:37]: i can update the doc string as 
`Optional last down threshold in hours to check if a bfd peer was down before those hours or not.`

File: anta/tests/bfd.py (hunk 5 above)
Comment (MaheshGSLAB) [reply anta/tests/bfd.py:60]: yes, earlier it was used in failure message so kept it same. Let me know if need to remove

File: anta/tests/bfd.py (hunk 1 above)
Comment (carl-baillargeon) [reply anta/tests/bfd.py:19]: Yes perfect

File: anta/tests/bfd.py (hunk 2 above)
Comment (carl-baillargeon) [reply anta/tests/bfd.py:37]: Maybe `down_threshold`?

File: anta/tests/bfd.py (hunk 3 above)
Comment (carl-baillargeon) [reply anta/tests/bfd.py:26]: Looks good, you can modify depending on how you decide to rename the `last_down` variable.

File: anta/tests/bfd.py (hunk 5 above)
Comment (carl-baillargeon) [reply anta/tests/bfd.py:60]: Can you investigate if the test can be refactored with a later revision (see my other comment) and check if this is still needed? Further testing might be required on EOS to check if this is necessary.

File: anta/tests/bfd.py (hunk 6 above)
Comment (MaheshGSLAB) [reply anta/tests/bfd.py:30]: In revision 1 we will get bfd peer IP and in revision 2 we will get the source address also. So for parsing the bfd details we need to go through from source address. Left one belong to revision 1 and right one belong to revision 2
Screenshot 2024-02-01 at 11 57 47

Revision 3 and 4 showing the same structure but if compare to revision 1 or 2 then we get an additional root keys as type of session. Agin left one is revision 1 and right one is revision 3.
Screenshot 2024-02-01 at 12 00 22

So i think revision 1 is the good one to parse the bfd details.

File: anta/tests/bfd.py (hunk 5 above)
Comment (MaheshGSLAB) [reply anta/tests/bfd.py:60]: As discussed removed the l3intf details from failure message.

File: tests/units/anta_tests/test_bfd.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+            },
+            {
+                "utcTime": 1703667348.111288,
+            },
+        ],
+        "inputs": {"down_threshold": 2},
+        "expected": {
+            "result": "failure",
+            "messages": [
+                "Following BFD peers were down:\n192.0.255.7 in default VRF was down 3 hours ago.\n"
```
Comment (carl-baillargeon) [inline tests/units/anta_tests/test_bfd.py:243]: Why it says the peer was down 3 hours ago when the provided threshold is 2?

Comment (MaheshGSLAB) [reply tests/units/anta_tests/test_bfd.py:243]: Fixed... Earlier i was checking that if peer was down for more then threshold hours then failing. Now changed it to if peer is down within threshold hours then fail it. If threshold is 2 hours and and peer was down 1 hour ago then test will fail. If peer was down 3 hours ago then it wont fail.

**Review by carl-baillargeon — APPROVED** [review]

**Comment by carl-baillargeon** [comment]
LGTM

**Pull request merged**